pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package backupsink
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/backup/backuppb"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/admission"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+type SSTSinkKeyWriter struct {
+	FileSSTSink
+	// prevKey represents the last key written using WriteKey. When writing a new
+	// key, this helps determine if the last key written was mid-row. This resets
+	// on each flush.
+	prevKey roachpb.Key
+}
+
+func MakeSSTSinkKeyWriter(
+	conf SSTSinkConf, dest cloud.ExternalStorage, pacer *admission.Pacer,
+) *SSTSinkKeyWriter {
+	return &SSTSinkKeyWriter{
+		FileSSTSink: *MakeFileSSTSink(conf, dest, pacer),
+	}
+}
+
+// WriteKey writes a single key to the SST file. The key should be the full key.
+// span refers to the span that the key belongs to. start and end are the time
+// bounds of the span being backed up. The writing of each key could
+// potentially split up the span of the previously written key (if both keys
+// were part of the same span and the span had to be split). As a consequence
+// the span that is recorded for the new key is returned, as it may be a subspan
+// of the span passed in for the key.
+//
+// flush should be called after the last key is written to ensure that the SST
+// is written to the destination.
+func (s *SSTSinkKeyWriter) WriteKey(
+	ctx context.Context,
+	key storage.MVCCKey,
+	value []byte,
+	span roachpb.Span,
```

Comment:
I don't understand why the caller needs to pass `span`, `startTime` and `endTime` to `WriteKey()`. I think all of these can be inferred by the stream of in order, single version, point keys the caller passes to the sink key writer. Perhaps we should chat about this on Thursday.

Reply:
I don't think we necessarily know what the start and end time are from the stream of keys, do we? If I have we have a full backup taken at `t=2` and it contains an MVCC key from `t=1`, the compacted backup still should have a start of `t=2`. And I don't think end time can be extrapolated from the stream of keys, especially if you have multiple spans that you are writing that should be written to the same `BackupManifest_File`, in which case the start and end should be shared. We could create a wrapper `SSTSinkKeyWriterConf` that wraps `SSTSinkConf` where you set the start and end there.

w.r.t to `span`, I do think that is necessary. If we write one span and then switch to another span, that information is needed to determine whether the `BackupManifest_File` can be contiguously extended or a new one needs to be created. In the case of the latter (and when we flush), we need to know the `span.EndKey` so that we can determine if the last key written was written mid-row.

Reply:
on start and end time: in regular old backups, they are only ever set [here](https://github.com/kev-cao/cockroach/blob/backupccl/fileSSTSink-writeKey/pkg/backup/backup_processor.go#L682), when dealing with introduced spans (start time is nil instead of inc backup start time) and revision history export request resumptions. Also note they are never read during restore -- just used in the og file sink logic and checkpointing for backup resumptions. I don't think either use case applies for backup compactions, so do we ever need to set them?

for the span arg: i think what you're referring to is if we're in the following situation, right? We just finished processing restoreSpanEntry for key span (a-d], and we're about to begin processing restore span entry with key span (d-f]. We can extend manifest file as these two spans are adjacent. Could we instead check this on an `sstSink.Reset()` func at the top of the iteration loop? If they are not adjacent or the current manifest file entry is  big enough, then Reset() calls flush, else the current manifest file entry is extended. I don't see why this logic is required every time we write a key, nor how the mid-row stuff factors in here.

Reply:
Oh I see, in that case we don't have to worry about start and end time at all?

---

Yep that's the situation that I am describing. With a `sstSink.Reset()`, we would make an assertion that every time you call `sstSink.WriteKey` between calling `sstSink.Reset()`, you are writing keys that belong to the same span. Then when you call `.Reset()`, you would need the span you were previously writing as well as the span you are about to write. In order to facilitate that, we would need some way of caching the last span written, which ideally that logic is contained within the key writer, so we would want something like `sstSink.SetSpan()`.

So we'd have receiver functions:
```golang
func (s *SSTSinkKeyWriter) SetSpan(s roachpb.Span)

func (s *SSTSinkKeyWriter) Reset(nextSpan roachpb.Span) // Maybe naming it "NextSpan" or "AdvanceSpan" would be better?
```

Reply:
I think that sounds good to me, we would just need to document the situations in which these need to be called, and add some assertions as well (basically just checking that `SSTSinkKeyWriter.span` is set)

Reply:
It does split up the extension logic. `WriteKey` would check extension in the case of whether or not the `fileSpanByteLimit` was reached. `Reset()` would check extension in the case of whether or not the next span should extend the previous span.

Reply:
before implementing, let's chat through it on thursday